src/mapping/attractor_mapping_recurrences.jl

```diff
@@ -158,7 +158,7 @@ basins, making the computation faster as the grid is processed more and more.
     [Chaos 32, 023104 (2022)](https://doi.org/10.1063/5.0076568)
 """
 function basins_of_attraction(mapper::AttractorsViaRecurrences; show_progress = true)
-    basins = mapper.bsn_nfo.basins
+    basins = zeros(Int32, size(mapper.bsn_nfo.basins))
```

Comment:
I don't see how this would fix anything; cells that weren't processed before for whatever reason, still wouldn't be processed...

Reply:
It does fix the problems with the zeros. The problem in the new code is that the matrix `mapper.bsn_nfo.basins` is not updated with the basins value anywhere. The function `_get_label_ic!` returns the value but does not update the structure. It keeps only attractors for matching. 

Matching works perfectly but the possible speed improvement due to the basins identification is lost. It is a easy fix but I want to check if there is no regression before patching.

Reply:
I have tested the code including an update of the structure `mapper.bsn_nfo_basins` with the values of the basins (as for now it only keeps the attractors in the matrix). I do not recommend to update the matrix because: 
1. It slows everything  for your example. It is 10 times slower with the update of the matrix. I think the reason is that when the finite state machine enters the `bas_hit:` state it resets the counters and the method stumbles. 
2. It can be less precise. Results are different for the example you gave. Some initial conditions are labelled differently because the trajectory may come close to other basins. When the matrix is not updated there is no shortcut and you trust the matching only when you get to the attractor. 

If we want to keep this option, the code has to be modified slightly and I suggest a keyword to switch on this option that should be off by default. If we don't find it compelling, we can remove the state `bas_hit:` from the FSM and it will work just fine.

Reply:
Okay, I see why the basin hitting clause can slow things down here since we have large multistability. I agree with you for doing a PR that adds a keyword that turns off the basin hitting clause. But we can't remove it from the source code that's a breaking change!

Reply:
Sorry, matching in this context is pairing an initial condition to an attractor.

Reply:
I still haven't understood though how this simple change fixes the problem #73 has with the zeros. We haven't solved the actual problem yet, we have just completely excluded some parts of the source code from running.

Reply:
I think the problem with  `basins = mapper.bsn_nfo.basins` is that the matrix is not deep-copied. I am not expert on object reference in Julia but it seems that the two objects are not in independent in memory space. While I was debugging I saw some entries in the matrix basins where changing values. 

While if you assign a new matrix and update it entries, it works fine. Does it makes sense?

Reply:
But that is eaxctlu the point; there is no reason to deepcopy the matrix because the Finite STate Macine takes care of putting the correct values to a given cell uisng the `get_label_ic!(` function. Right?

Reply:
>  the Finite STate Macine takes care of putting the correct values to a given cell uisng the `get_label_ic!(` function. Right?

Not anymore, the mapper function returns the value but does not update the internal matrix value. The reason is that you can call any initial condition you want and the value is not tied to the internal grid of the mapper. I give you an example: 

* Suppose you have defined a grid for the recurrence of 100x100 in the box [-1,1]x[-1,1]. With the old version you could have only access to these 100x100 points in this box. With the new version you can call the mapper at **any coordinates**  in the box and it will give you the corresponding attractor. It is not cached in the internal matrix. 

In a sense you loose some information of previous computed initial conditions. But the gain in flexibility and precision is baffling.

Reply:
yes sure but the update of the basins happens later in this code when we d0 `basins[ind] = get_label_ic!(`. and that's why we want to use the same matrix the mapper stores internally, so that hte internal FSM also has access to cells labelled as basins and can also use the basin convergence clause.

Reply:
Found the bug. Tt is a book keeping error. When the function returns -1 we have to make sure that the temporary labelled cells returns to zero. 

This fix the code for now. But I want to check if there is a better way to do this.